src/IO.cpp

```diff
@@ -5,6 +5,10 @@
 #include <fstream>
 #include <sstream>
 
+#ifndef ROOT_DIR
```

Comment:
This block of code should only be in one place, not both here and in `main.cpp`. I'd suggest moving it to `IO.h`, as then it will be picked up by any source file that is making use of I/O (such as `main.cpp`).

Also, I would replace this with:
```
#ifndef ROOT_DIR
#error Macro ROOT_DIR must be defined!
#endif
```
As it is not safe to build the code without defining this macro. The `#error` directive will terminate the build and print the error message.

Reply:
Changes should (hopefully) be present now, thanks for that, I am learning!